### ISSUE TITLE
What's the purpose of the torch.distributed.barrier()?

### ISSUE THREAD

**Issue by Zarca** [dmlc/dgl#1892]
Hi ,
I'm wondering what's the intention of the torch.distributed.barrier() called in the example of graphsage multi gpu trainig? as showed the link below:
https://github.com/dmlc/dgl/blob/68a978d4de27b6625870e1ea4242075e990922f4/examples/pytorch/graphsage/train_sampling_multi_gpu.py#L213

For my part , i suppose this line should be put at the line **between Line 202( loss.backward() ) and line 203 (optimizer.step())** :
https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/train_sampling_multi_gpu.py#L202

to ensure a synchronization on the gradient for all processes. 

Thanks in advance and looking forward your help XD

**Comment by yzh119** [comment]
@Zarca , it's not necessary to put `barrier` between line 202 and 203 because pytorch inherently synchronizes all processes (PyTorch distributed uses AllReduce algorithm) during parameters update.

We put a barrier in line 213 to enforce all processes to synchronize before evaluation phase, so that the time meatured in line 215 is the same for all processes. It's not very important considering the program is already synchronized over the backward process.

**Comment by Zarca** [comment]
@ yzh119 Hello yzh119, I see, yep, it make sense to keep all processes tic tock time the same.

Still the very same example , I'm also wondering why you use the multiprocessing(mp).Process class to fork child process to do 
jobs about cuda tensor ln line 283:
https://github.com/dmlc/dgl/blob/68a978d4de27b6625870e1ea4242075e990922f4/examples/pytorch/graphsage/train_sampling_multi_gpu.py#L283
However, I thought the official torch document said we must use mp.spawn() function to do the multi-gpu training ,  otherwise there will be a error like this while i switch another dataset to run the very same script:
![image](https://user-images.githubusercontent.com/24642659/88914027-8f561e80-d294-11ea-9658-57a16b52c4aa.png)
Have you ever encounterd such error before ?

**Comment by VoVAllen** [comment]
@Zarca You are right, using fork with cuda context is tricky. That you need to ensure the context is initialized after forking the child process.
For the spawn method, you can use it with the latest nightly build (by pip install --upgrade --pre dgl-cuXX). (Previous version you may meet bugs)

**Comment by Zarca** [comment]
@VoVAllen Oh, I just figure it out accroding to what you suggested, that is to initial cuda after the fork() , 
thanks a lot ! XD
